### ISSUE TITLE
Using AS3 with GSLB_Server and virtualServerDiscoveryMode=enabled-no-delete does not allow GSLB_Pool to reference the virtual servers

### ISSUE THREAD

**Issue by dwebr** [F5Networks/f5-appsvcs-extension#578]
### Environment
 * Application Services Version: 3.34.0
 * BIG-IP Version: BIG-IP 16.1.0 Build 0.0.19 Final 

### Summary
When using AS3 with GSLB_Server that has "virtualServerDiscoveryMode": "enabled-no-delete", the virtual servers from the LTM get discovered and when the GSLB_Pool is declared the virtual server can be referenced by name and the POST is successful. However, the  GSLB_Pool is left in "Offline (Enabled) - No enabled pool members are available" state with no members added to the pool. The members can be added using the configuration GUI and the members and pool with go to an up status,

### Steps To Reproduce
1. Add LTM virtual servers
2. GTM/LTM iquery in working state
3. Submit the following declarations for tenant /Common and then tenant /Core:

Tenant /Common/Shared declaration
```json
{
    "class": "AS3",
    "action": "deploy",    
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.34.0",
        "id": "GTM_COMMON",
        "Common": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "PD-Core-DC": {
                    "class": "GSLB_Data_Center"
                },
                "PD-Core-CL1x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CL1x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.69",
                            "label": "PD-L4D-F5VA-CL11"
                        },
                        {
                            "address": "172.29.128.70",
                            "label": "PD-L4D-F5VA-CL12"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                },
                "PD-Core-CL2x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CL2x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.73",
                            "label": "PD-L4D-F5VA-CL21"
                        },
                        {
                            "address": "172.29.128.74",
                            "label": "PD-L4D-F5VA-CL22"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                },
                "PD-Core-CD1x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CD1x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.77",
                            "label": "PD-L4D-F5VA-CD11"
                        },
                        {
                            "address": "172.29.128.78",
                            "label": "PD-L4D-F5VA-CD12"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                }
            }
        }
    }
}
```
Tenant /Core declaration
```json
{
    "class": "AS3",
    "action": "deploy",
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.34.0",
        "id": "GTM_CORE",
        "Core": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "web01.core.lab.thrivent.com": {
                    "class": "GSLB_Domain",
                    "resourceRecordType": "A",
                    "domainName": "web01.core.lab.thrivent.com",
                    "pools": [
                        {
                            "use": "C3D1-031-WEB01-EPO-T8080"
                        }
                    ]
                },
                "C3D1-031-WEB01-EPO-T8080": {
                    "class": "GSLB_Pool",
                    "resourceRecordType": "A",
                    "lbModePreferred": "round-robin",
                    "members": [
                        {
                            "ratio": 1,
                            "server": {
                                "use": "/Common/Shared/PD-Core-CL1x"
                            },
                            "virtualServer": "/Common/AC3D1-031-WEB01-EVS-T8080",
                            "dependsOn": "none"
                        }
                    ]
                }
            }
        }
    }
}
```
POST RESULTS
```json
{
    "id": "c759ed24-81e9-4b49-af92-68161f78ffb1",
    "results": [
        {
            "code": 200,
            "message": "success",
            "lineCount": 16,
            "host": "localhost",
            "tenant": "Common",
            "runTime": 1791
        },
        {
            "code": 200,
            "message": "success",
            "lineCount": 18,
            "host": "localhost",
            "tenant": "Core",
            "runTime": 1438
        },
        {
            "code": 200,
            "message": "success",
            "lineCount": 20,
            "host": "localhost",
            "tenant": "Common",
            "runTime": 1811
        }
    ],
    "declaration": {
        "Core": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "web01.core.lab.thrivent.com": {
                    "class": "GSLB_Domain",
                    "resourceRecordType": "A",
                    "domainName": "web01.core.lab.thrivent.com",
                    "pools": [
                        {
                            "use": "C3D1-031-WEB01-EPO-T8080"
                        }
                    ]
                },
                "C3D1-031-WEB01-EPO-T8080": {
                    "class": "GSLB_Pool",
                    "resourceRecordType": "A",
                    "lbModePreferred": "round-robin",
                    "members": [
                        {
                            "ratio": 1,
                            "server": {
                                "use": "/Common/Shared/PD-Core-CL1x"
                            },
                            "virtualServer": "/Common/AC3D1-031-WEB01-EVS-T8080",
                            "dependsOn": "none"
                        }
                    ]
                }
            }
        },
        "Common": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "PD-Core-DC": {
                    "class": "GSLB_Data_Center"
                },
                "PD-Core-CL1x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CL1x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.69",
                            "label": "PD-L4D-F5VA-CL11"
                        },
                        {
                            "address": "172.29.128.70",
                            "label": "PD-L4D-F5VA-CL12"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                },
                "PD-Core-CL2x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CL2x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.73",
                            "label": "PD-L4D-F5VA-CL21"
                        },
                        {
                            "address": "172.29.128.74",
                            "label": "PD-L4D-F5VA-CL22"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                },
                "PD-Core-CD1x": {
                    "class": "GSLB_Server",
                    "label": "PD-Core-CD1x-Server",
                    "serverType": "bigip",
                    "dataCenter": {
                        "use": "PD-Core-DC"
                    },
                    "devices": [
                        {
                            "address": "172.29.128.77",
                            "label": "PD-L4D-F5VA-CD11"
                        },
                        {
                            "address": "172.29.128.78",
                            "label": "PD-L4D-F5VA-CD12"
                        }
                    ],
                    "virtualServerDiscoveryMode": "enabled-no-delete",
                    "exposeRouteDomainsEnabled": true
                }
            }
        },
        "class": "ADC",
        "schemaVersion": "3.34.0",
        "id": "GTM_CORE",
        "controls": {
            "class": "Controls",
            "userAgent": "vscode-f5/3.5.0",
            "archiveTimestamp": "2022-03-16T01:37:13.816Z"
        },
        "updateMode": "selective"
    }
}
```
4. Observe the following error response:
The screenshots show the Pool as offline with no available pool members, yet a member that was discovered from the LTM can be added manually.
![image](https://user-images.githubusercontent.com/95582257/158498755-453dd93e-e8f3-495a-9a08-1323f5a9447e.png)
![image](https://user-images.githubusercontent.com/95582257/158498781-16322d5a-5055-427d-94a6-57e3e66b6dd0.png)

TMSH Running Configuration
```
gtm server PD-Core-CL1x {
    addresses {
        172.29.128.69 {
            device-name 0
        }
        172.29.128.70 {
            device-name 1
        }
    }
    datacenter PD-Core-DC
    devices {
        0 {
            addresses {
                172.29.128.69 { }
            }
        }
        1 {
            addresses {
                172.29.128.70 { }
            }
        }
    }
    expose-route-domains yes
    metadata {
        as3 { }
    }
    monitor bigip
    product bigip
    virtual-server-discovery enabled-no-delete
    virtual-servers {
        /Common/AC3D1-031-WEB01-EVS-T8080 {
            destination 172.29.144.31:http
        }
        /Common/AC3D1-032-WEB02-EVS-T8081 {
            destination 172.29.144.32:http
        }
        /Common/AC3D1-033-WEB03-EVS-T8083 {
            destination 172.29.144.33:http
        }
        /Common/AC3D1-034-WEB04-EVS-T8084 {
            destination 172.29.144.34:http
        }
    }
}
```

Pool member after AS3 POST.
```
gtm pool a C3D1-031-WEB01-EPO-T8080 {
    partition Core
}
```

Pool after adding the virtual server manually via the configuration GUI.
```
gtm pool a C3D1-031-WEB01-EPO-T8080 {
    members {
        /Common/PD-Core-CL1x:/Common/AC3D1-031-WEB01-EVS-T8080 {
            member-order 0
        }
    }
    partition Core
}
```
### Expected Behavior
The reference for virtualServer in the GSLB_Pool member should correctly reference the GSLB_Server virtual server that was discovered through Virtual Server Discovery.

### Actual Behavior
The GSLB_Pool is configured but no pool members are added even though the declaration is valid and accepted by AS3.

**Comment by dwebr** [comment]
It appears that using the "bigip" property instead of "use" when defining the pool member server allows for the successful tie to the discovered virtual server. This seems a little odd, since the GSLB_Server was defined with AS3, albeit in a different declaration file. 

My understanding was that "**use** indicates a reference to another class object in the declaration. **bigip** indicates a component pathname to an object that was created outside of AS3 (typically in /Common)." 

Is this the correct use of the "bigip" property clause? Is it required since the virtual servers associated with the "server" were not defined with AS3?

```json
{
    "class": "AS3",
    "action": "deploy",
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.34.0",
        "id": "GTM_CORE",
        "Core": {
            "class": "Tenant",
            "Shared": {
                "class": "Application",
                "template": "shared",
                "web01.core.lab.thrivent.com": {
                    "class": "GSLB_Domain",
                    "resourceRecordType": "A",
                    "domainName": "web01.core.lab.thrivent.com",
                    "pools": [
                        {
                            "use": "C3D1-031-WEB01-EPO-T8080"
                        }
                    ]
                },
                "C3D1-031-WEB01-EPO-T8080": {
                    "class": "GSLB_Pool",
                    "resourceRecordType": "A",
                    "lbModePreferred": "round-robin",
                    "members": [
                        {
                            "ratio": 1,
                            "server": {
                                "bigip": "/Common/PD-Core-CL1x"
                            },
                            "virtualServer": "/Common/AC3D1-031-WEB01-EVS-T8080",
                            "dependsOn": "none"
                        }
                    ]
                }
            }
        }
    }
}
```

**Comment by dstokesf5** [comment]
Thank you for your feedback. I have added this issue to our internal product backlog as AUTOTOOL-3066.

**Comment by antonywm** [comment]
I also have this same issue in that when GSLB_Server is defined with auto discovery of Virtual Servers (either with enabled or enabled-no-delete) and GSLB_Pool is in the same declaration, the declaration seems to succeed with a 200 OK but no virtual servers are in the pools.

In the end the only way I could get it to work is to send 2 declarations - firstly to build the server (in Common because you have to), and then separately to build the WIP and the Pools (which I am building in a partition). This seems to work. I wondered if there is some delay issue - where it takes time for the auto-discovery process to populate the virtual servers. However, this would not explain why the validation works - you'd expect it to fail saying the virtual server did not exist.

**Comment by antonywm** [comment]
sorry - should also mention... in the second declaration containing the GSLB_Pool objects, I have to refer to the server using the "bigip" property (as they already exist, having been defined in the first declaration) - Example:

```json
"members": [
   {
      "server": {
	   "bigip": "/Common/myServer"
	   },
      "virtualServer": "/myPartition/dev/myVirtualServer"
      }
   ],
```

**Comment by dstokesf5** [comment]
@antonywm Yes, I suspect this is a timing problem.

@dwebr, @antonywm The "use" property should be appropriate here since you are referring to an object defined by AS3.

**Comment by seattlevine** [comment]
There are a couple issues here. 
- One is timing because the gtm sub-system takes some time to discover ltm virtuals when `virtualServerDiscoveryMode` is enabled. We are resolving this by adding a delay before creating any gtm pools that refer to an existing non-AS3 gtm server or refer to an AS3 gtm server that has `virtualServerDiscoveryMode` enabled.
- `use` is the correct property, however there is a bug here with a combination of `use` and `virtualServerDiscoveryMode`. We are resolving this issue as well.

**Comment by sunitharonan** [comment]
This has been resolved in the 3.40.0 release of AS3.
